File: NOVEMBER-2025/23-11-2025/Maximum Stone Removal.cpp
class Solution {
public:
    unordered_map<int, int> parent;
    int find(int x) {
        if(parent.find(x) == parent.end()) parent[x] = x;
        if(parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    void unite(int x, int y) {
        int px = find(x);
        int py = find(y);
        if(px != py) parent[px] = py;
    }
    int maxRemove(vector<vector<int>>& stones) {
        for(auto &stone : stones) {
            unite(stone[0], stone[1] + 10001);
        }
        unordered_set<int> components;
        for(auto &stone : stones) {
            components.insert(find(stone[0]));
        }
        return stones.size() - components.size();
    }
};

File: NOVEMBER-2025/23-11-2025/Maximum Stone Removal.java
import java.util.*;

class Solution {
    Map<Integer, Integer> parent = new HashMap<>();
    
    private int find(int x) {
        if (!parent.containsKey(x)) parent.put(x, x);
        if (parent.get(x) != x) parent.put(x, find(parent.get(x)));
        return parent.get(x);
    }
    
    private void union(int x, int y) {
        int px = find(x);
        int py = find(y);
        if (px != py) parent.put(px, py);
    }
    
    public int maxRemove(int[][] stones) {
        for (int[] stone : stones) {
            union(stone[0], stone[1] + 10001);
        }
        
        Set<Integer> components = new HashSet<>();
        for (int[] stone : stones) {
            components.add(find(stone[0]));
        }
        
        return stones.length - components.size();
    }
}